src/components/controller.cpp

```diff
@@ -76,42 +76,33 @@ controller::controller(connection& conn, signal_emitter& emitter, const logger&
 
   m_log.trace("controller: Setup user-defined modules");
   size_t created_modules{0};
+  created_modules += setup_modules(alignment::LEFT);
+  created_modules += setup_modules(alignment::CENTER);
+  created_modules += setup_modules(alignment::RIGHT);
 
-  for (int i = 0; i < 3; i++) {
-    alignment align{static_cast<alignment>(i + 1)};
-    string configured_modules;
-
-    if (align == alignment::LEFT) {
-      configured_modules = m_conf.get(m_conf.section(), "modules-left", ""s);
-    } else if (align == alignment::CENTER) {
-      configured_modules = m_conf.get(m_conf.section(), "modules-center", ""s);
-    } else if (align == alignment::RIGHT) {
-      configured_modules = m_conf.get(m_conf.section(), "modules-right", ""s);
-    }
-
-    for (auto& module_name : string_util::split(configured_modules, ' ')) {
-      if (module_name.empty()) {
-        continue;
-      }
-
-      try {
-        auto type = m_conf.get("module/" + module_name, "type");
-
-        if (type == "custom/ipc" && !m_ipc) {
-          throw application_error("Inter-process messaging needs to be enabled");
-        }
+  if (!created_modules) {
+    throw application_error("No modules created");
+  }
 
-        m_modules[align].emplace_back(make_module(move(type), m_bar->settings(), module_name, m_log));
-        created_modules++;
-      } catch (const runtime_error& err) {
-        m_log.err("Disabling module \"%s\" (reason: %s)", module_name, err.what());
+  /*
+   * Check if each module name only appears once
+   * Iterate over each module and check if its name appears anywhere else
+   */
+  for (auto it = m_modules.begin(); it != m_modules.end(); it++) {
```

Comment:
This loop prints the error log X times if the module appears X times (with X > 2).
If you really want to really address this problem you can replace the whole loop with something like this:
```cpp
  std::sort(m_modules.begin(), m_modules.end(), [](const auto& m1, const auto& m2) { return m1->name() < m2->name(); });

  auto dup_it = m_modules.cbegin();

  do {
  	auto equal_predicate = [](const auto& m1, const auto& m2) { return m1->name() == m2->name(); };
    dup_it = std::adjacent_find(dup_it, m_modules.cend(), equal_predicate);

    if (dup_it != m_modules.cend()) {
	    auto first_mismatch = std::mismatch(dup_it, m_modules.cend(), dup_it, equal_predicate).first;
	    m_log.err(
          "The module \"%s\" appears %d times in your modules list. "
          "This should be avoided, as it can lead to inconsistent behavior. "
          "Both modules will be displayed for now.",
          (*dup_it)->name(), std::distance(dup_it, first_mismatch));

	    dup_it = first_mismatch;
    } else {
    	break;
    }
  } while (true);
```

Reply:
Maybe the message should state this is deprecated, even if the error already indicates the problem.

Reply:
This doesn't print an error message for all modules that appear multiple times though. I think you have a bug in there somewhere.

I agree, I'll change the error message.

Reply:
> This doesn't print an error message for all modules that appear multiple times though. I think you have a bug in there somewhere.

Could you give me an example when it doesn't work? I'll post a corrected version.

Reply:
```
[bar/example]
modules-left = a b c
modules-center = a b
modules-right = b a
```

It says `a` appears 7 times.

Reply:
Thanks !

Reply:
I found a fix.